lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,2561 @@
+    else
+      break;
+  }
+  return operand_idx;
+}
+
+class LifeTime {
+protected:
+  uint begin;
+  uint end;
+
+public:
+  LifeTime(uint begin, uint end) : begin(begin), end(end) {
+    assert(end >= begin && "invalid lifetime");
+  };
+
+  bool isAfter(LifeTime *other) { return begin > other->end; }
+
+  bool isOverlapping(LifeTime *other) {
+    return !isAfter(other) && !other->isAfter(this);
+  }
+
+  uint distance(LifeTime *other) {
+    if (isOverlapping(other))
+      return 0;
+    return std::max(begin, other->begin) - std::min(end, other->end);
+  }
+
+  void combine(LifeTime *other) {
+    begin = std::min(begin, other->begin);
+    end = std::max(end, other->end);
+  }
+
+  uint getBegin() { return begin; }
+  uint getEnd() { return end; }
+};
+
+class LifeTimeAnalysis {
+private:
+  StringRef name;
+  SmallVector<LifeTime *> lifetimes;
+  SetVector<PhysicalQID> frame;
+  // DenseMap<VirtualQID, PhysicalQID> allocMap;
+
+  /// Given a candidate lifetime, tries to find a qubit to reuse,
+  /// and otherwise allocates a new qubit
+  PhysicalQID allocatePhysical(LifeTime *lifetime) {
+    std::optional<size_t> best_reuse = std::nullopt;
+    std::optional<size_t> empty = std::nullopt;
+    uint best_distance = INT_MAX;
+
+    for (uint i = 0; i < lifetimes.size(); i++) {
+      if (!lifetimes[i]) {
+        empty = i;
+        continue;
+      }
+
+      auto other = lifetimes[i];
+      auto distance = lifetime->distance(other);
+      if (!lifetime->isOverlapping(other) && distance < best_distance) {
+        best_reuse = i;
+        best_distance = distance;
+      }
+    }
+
+    // Reuse a qubit based on its lifetime in the same scope
+    if (best_reuse) {
+      auto physical = best_reuse.value();
+      lifetimes[physical]->combine(lifetime);
+      delete lifetime;
+      return physical;
+    }
+
+    // Reuse a qubit without a lifetime (used in a different frame)
+    if (empty) {
+      auto physical = empty.value();
+      lifetimes[physical] = lifetime;
+      return physical;
+    }
+
+    // Fall back: allocate a new qubit
+    lifetimes.push_back(lifetime);
+    return lifetimes.size() - 1;
+  }
+
+public:
+  LifeTimeAnalysis(StringRef name) : name(name), lifetimes(), frame() {}
+
+  ~LifeTimeAnalysis() {
+    for (auto lifetime : lifetimes)
+      if (lifetime)
+        delete lifetime;
+  }
+
+  PhysicalQID allocatePhysical(VirtualQID qid, LifeTime *lifetime) {
+    auto phys = allocatePhysical(lifetime);
+    frame.insert(phys);
+    return phys;
+  }
+
+  SetVector<PhysicalQID> getAllocated() {
+    for (uint i = 0; i < lifetimes.size(); i++) {
+      if (lifetimes[i])
+        delete lifetimes[i];
+      lifetimes[i] = nullptr;
+    }
+    auto pqids = SetVector<PhysicalQID>(frame);
+    frame.clear();
+    return pqids;
+  }
+
+  void reallocatePhysical(PhysicalQID phys, LifeTime *lifetime) {
+    lifetimes[phys] = lifetime;
+  }
+
+  size_t getCount() { return lifetimes.size(); }
+
+  void print() {
```

Comment:
```suggestion
  void dump() {
```
LLVM uses `dump` everywhere.

Reply:
Fixed